commons/protocol/src/main/scala/sbt/protocol/Protocol.scala

```diff
@@ -217,54 +215,46 @@ final case class BackgroundJobLogEvent(jobId: Long, entry: LogEntry) extends Log
   require(jobId != 0L)
 }
 
-/** A custom event from a task. "name" is conventionally the simplified class name. */
-final case class TaskEvent(taskId: Long, name: String, serialized: SerializedValue) extends Event
+/** A custom event from a task. */
+final case class TaskEvent(taskId: Long, serialized: SerializedValue) extends Event
 
 object TaskEvent {
   def apply[T: SPickler](taskId: Long, event: T): TaskEvent = {
     val serialized = SerializedValue(event)
-    TaskEvent(taskId, MessageSerialization.makeSimpleName(event.getClass), serialized)
+    TaskEvent(taskId, serialized)
   }
 }
 
 /** Companion objects of events which can go in a task event extend this */
 trait TaskEventUnapply[T] {
-  import scala.reflect.ClassTag
-
-  def unapply(event: Event)(implicit unpickler: Unpickler[T], classTag: ClassTag[T]): Option[(Long, T)] = event match {
+  def unapply(event: Event)(implicit unpickler: Unpickler[T]): Option[(Long, T)] = event match {
     case taskEvent: TaskEvent =>
-      val name = MessageSerialization.makeSimpleName(implicitly[ClassTag[T]].runtimeClass)
-      if (name != taskEvent.name) {
+      if (taskEvent.serialized.hasTag[T])
+        Some(taskEvent.taskId -> taskEvent.serialized.parse[T].get)
```

Comment:
Unpickler.canUnpickleTag would be very simple to add to the macros I suppose - we already build a dispatch on tag so we just build the same dispatch but retuning true/false. By default it could do the comparison with Unpickler.tag for people writing manual Unpicklers....